### PR TITLE
Feature/wjiang/io restructure

### PR DESCRIPTION
**Pull request by weiyuan-jiang**
## Description

The existing codes will have to share the i_clients and o_clients across the caps or applications. This creates a problem when initializing the io clients and servers. In the new implementation, each application with an independent cap will have its own i_clients and o_clients  ( it is renamed as iclient_manager and oclient_manaer) which are stored in a map.  
 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:

- [x] I have tested this change with a run of GEOSgcm (if non-trivial)
- [x] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [ ] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

### CODE REVIEW

**Comment by weiyuan-jiang** [comment]
@bena-nasa I removed some functions in MAPL_CFIO that I thought was obsolete. Please review it when you get chances

**Comment by weiyuan-jiang** [comment]
Before the merge, I want a thorough test. I only tested on GEOSldas so far.

File: GMAO_pFIO/ClientManager.F90 [hunk 1]
```diff
@@ -398,6 +412,64 @@ function current(this) result(clientPtr)
       clientPtr=> this%clients%at(this%current_client)
    end function current
 
+   subroutine set_ideal_client(this,nwriting,unusable,rc)
```
Comment (tclune) [inline GMAO_pFIO/ClientManager.F90:415]: Document what "nwriting" corresponds to (data, fields, files).

File: GMAO_pFIO/ClientManager.F90 [hunk 2]
```diff
@@ -398,6 +412,64 @@ function current(this) result(clientPtr)
...
+   subroutine split_clients(this, nsplit, nwriteCutoff,unusable, rc)
+      class (ClientManager), intent(inout) :: this
+      integer, intent(in) :: nsplit
+      integer, intent(in) :: nWriteCutoff
+      class (KeywordEnforcer),  optional, intent(in) :: unusable
+      integer, optional, intent(out) :: rc
+
+      integer :: i
+
+      do i=1,this%server_sizes%size()
```
Comment (tclune) [inline GMAO_pFIO/ClientManager.F90:450]: Add guard logic to ensure this is only called once.

File: GMAO_pFIO/ClientManager.F90 [hunk 3]
```diff
@@ -398,6 +412,64 @@ function current(this) result(clientPtr)
...
+             call this%small_server_pool%push_back(i)
+          end if
+      enddo
+      this%writeCutoff = nwriteCutoff
+      _RETURN(_SUCCESS)
+      _UNUSED_DUMMY(unusable)
+
+   end subroutine split_clients
+
+   subroutine set_server_size(this, server_size, unusable, rc)
```
Comment (tclune) [inline GMAO_pFIO/ClientManager.F90:463]: Maybe the server_sizes vector should be in the same structure as the thread?

File: GMAO_pFIO/DirectoryService.F90 [hunk 4]
```diff
@@ -99,8 +102,8 @@ function new_DirectoryService(comm, unusable, rc) result(ds)
       ! Create windows that will be used for coordination
       ! 1. lock - control modification of other windows
       ds%mutex = MpiMutex(ds%comm)
-      ds%win_server_directory = make_directory_window(ds%comm)
-      ds%win_client_directory = make_directory_window(ds%comm)
+      ds%win_server_directory = make_directory_window(ds%comm,ds%server_dir)
```
Comment (tclune) [inline GMAO_pFIO/DirectoryService.F90:105]: either make a subroutine or a function that returns all needed info

Comment (tclune) [reply GMAO_pFIO/DirectoryService.F90:105]: Also "_dir" should be "_addr"

File: GMAO_pFIO/DirectoryService.F90 [hunk 5]
```diff
@@ -116,28 +119,37 @@ function new_DirectoryService(comm, unusable, rc) result(ds)
...
-      call c_f_pointer(addr, dir)
-      call MPI_Win_create(dir, sz, 1, MPI_INFO_NULL, comm, win, ierror)
+      if (rank == 0)  then
+         sz = sizeof_directory()
+         call MPI_Alloc_mem(sz, MPI_INFO_NULL, addr, ierror)
+         call c_f_pointer(addr, dir)
+         call MPI_Win_create(dir, sz, 1, MPI_INFO_NULL, comm, win, ierror)
+      else
+         sz = 0
+         call MPI_Win_create(dir_null, sz, 1, MPI_INFO_NULL, comm, win, ierror)
```
Comment (tclune) [inline GMAO_pFIO/DirectoryService.F90:140]: Move Win_create() outside of the if.   Just set arguments inside the if.

**Comment by tclune** [comment]
@weiyuan-jiang I'm reviewing open PR's in MAPL.   What was the final intent with this one?   I don't see any changes after my comments.     Maybe this is where I suggested we instead focus on improving the back end of the server?   If so, should this PR be closed?

**Comment by weiyuan-jiang** [comment]
OK, I close the PR for now. I will address your issue and keep focusing on back end server.

**Pull request closed without merging**